File: api/src/main/java/dev/obscuria/fragmentum/network/PayloadRegistrar.java
package dev.obscuria.fragmentum.network;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.player.Player;

import java.util.function.BiConsumer;
import java.util.function.Function;

@SuppressWarnings("unused")
public interface PayloadRegistrar
{
    void allowClientOnly();

    void allowServerOnly();

    <T> void registerClientbound(Class<T> clazz, BiConsumer<T, FriendlyByteBuf> encoder, Function<FriendlyByteBuf, T> decoder, BiConsumer<Player, T> handler);

    <T> void registerServerbound(Class<T> clazz, BiConsumer<T, FriendlyByteBuf> encoder, Function<FriendlyByteBuf, T> decoder, BiConsumer<ServerPlayer, T> handler);
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/network/PayloadRegistrar.kt
@file:Suppress("unused", "RedundantNullableReturnType")

package dev.obscuria.fragmentum.network

import net.minecraft.network.FriendlyByteBuf
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Player
import java.util.function.BiConsumer
import java.util.function.Function

interface PayloadRegistrar {

    fun allowClientOnly()

    fun allowServerOnly()

    fun <T : Any> registerClientbound(
        clazz: Class<T>,
        encoder: BiConsumer<T, FriendlyByteBuf>,
        decoder: Function<FriendlyByteBuf, T>,
        handler: BiConsumer<Player, T>
    )

    fun <T : Any> registerServerbound(
        clazz: Class<T>,
        encoder: BiConsumer<T, FriendlyByteBuf>,
        decoder: Function<FriendlyByteBuf, T>,
        handler: BiConsumer<ServerPlayer, T>
    )
}